Don't generate Overlay Regions for fixed elements used as backgrounds

etiennesegonzac opened the pull request:
#### ede4f9d2750ec410851690fd28c42ddee8de7140
<pre>
Don&apos;t generate Overlay Regions for fixed elements used as backgrounds
<a href="https://bugs.webkit.org/show_bug.cgi?id=282748">https://bugs.webkit.org/show_bug.cgi?id=282748</a>
&lt;<a href="https://rdar.apple.com/135479679">rdar://135479679</a>&gt;

Reviewed by Mike Wyrzykowski.

Maintain a separate `scrollOverlayRegion` on EventRegion so we can
ignore some elements. Start with skipping fixed elements used as
background.

* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegion::EventRegion):
(WebCore::EventRegion::unite):
(WebCore::EventRegion::uniteScrollOverlayRegion):
(WebCore::EventRegion::scrollOverlayRegion const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
Introduce the new `scrollOverlayRegion`.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(addOverlayEventRegions):
Use the new `EventRegion#scrollOverlayRegion` when generating Overlay Regions.

* LayoutTests/overlay-region/overlay-background-expected.txt: Added.
* LayoutTests/overlay-region/overlay-background.html: Added.
Add a test.

Canonical link: <a href="https://commits.webkit.org/286390@main">https://commits.webkit.org/286390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d34b358a229316dc75bd805a61a04a854e81675

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75877 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54906 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28774 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80374 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27143 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77993 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64048 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3200 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59493 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17653 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78944 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/49369 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65167 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39853 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46769 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22650 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25470 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67884 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22988 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81838 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3246 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2043 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67723 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3400 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65135 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67030 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16705 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10973 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9101 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3196 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6002 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3217 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4155 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3224 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -407,21 +407,27 @@ EventRegion::EventRegion(Region&& region
     : m_region(WTFMove(region))
 #if ENABLE(TOUCH_ACTION_REGIONS)
-    , m_touchActionRegions(WTFMove(touchActionRegions))
+        , m_touchActionRegions(WTFMove(touchActionRegions))
```
super minor but I think these indents should be reverted

mwyrzykowski reviewed (approved):
Seems fine with the indentation fixed

etiennesegonzac committed:
Don't generate Overlay Regions for fixed elements used as backgrounds
https://bugs.webkit.org/show_bug.cgi?id=282748
<rdar://135479679>

Reviewed by Mike Wyrzykowski.

Maintain a separate `scrollOverlayRegion` on EventRegion so we can
ignore some elements. Start with skipping fixed elements used as
background.

* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegion::EventRegion):
(WebCore::EventRegion::unite):
(WebCore::EventRegion::uniteScrollOverlayRegion):
(WebCore::EventRegion::scrollOverlayRegion const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
Introduce the new `scrollOverlayRegion`.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(addOverlayEventRegions):
Use the new `EventRegion#scrollOverlayRegion` when generating Overlay Regions.

* LayoutTests/overlay-region/overlay-background-expected.txt: Added.
* LayoutTests/overlay-region/overlay-background.html: Added.
Add a test.

Canonical link: https://commits.webkit.org/286390@main
